src/cargo/util/rustc.rs

```diff
@@ -42,11 +44,28 @@ impl Rustc {
             triple.to_string()
         };
 
+        let mut cmd = util::process(&path);
```

Comment:
So, I experimented a bit. I think I can fairly easily change `probe_target_info` to have nice behaviour, but at the expense of messaging to the user. So, we can `cargo clean` without an error, with or without --crate-type metadata in the compiler, but if the compiler doesn't support it, then we give a warning about not supporting cargo check (even though the user just did a clean, not a check). Then if you do try to do `cargo check` with a compiler without metadata crates, then you get the same warning, but then Cargo bails out with an unfriendly internal error (I guess we could make that friendlier), rather than bailing out with the appropriate error (because now it is only a warning).

And of course we don't need to do anything to feature detect the crate type.

Does that sound like a good compromise?

Otherwise, there is always the hack of not collecting rmeta files to delete and deleting an rmeta for each rlib.